realm/src/main/java/io/realm/BaseRealm.java

```diff
@@ -436,19 +436,21 @@ protected synchronized void acquireFileReference(RealmConfiguration configuratio
      * Releases a reference to the Realm file. If reference count reaches 0 any cached configurations
      * will be removed.
      */
-    protected synchronized void releaseFileReference() {
-        String canonicalPath = configuration.getPath();
-        List<RealmConfiguration> pathConfigurationCache = globalPathConfigurationCache.get(canonicalPath);
-        pathConfigurationCache.remove(configuration);
-        if (pathConfigurationCache.isEmpty()) {
-            globalPathConfigurationCache.remove(canonicalPath);
-        }
+    protected void releaseFileReference() {
+        synchronized (BaseRealm.class) {
+            String canonicalPath = configuration.getPath();
+            List<RealmConfiguration> pathConfigurationCache = globalPathConfigurationCache.get(canonicalPath);
+            pathConfigurationCache.remove(configuration);
+            if (pathConfigurationCache.isEmpty()) {
+                globalPathConfigurationCache.remove(canonicalPath);
+            }
 
-        Integer refCount = globalRealmFileReferenceCounter.get(canonicalPath);
-        if (refCount == null || refCount == 0) {
-            throw new IllegalStateException("Trying to release a Realm file that is already closed");
+            Integer refCount = globalRealmFileReferenceCounter.get(canonicalPath);
```

Comment:
there is also this with `globalRealmFileReferenceCounter` since it's protected in `acquireFileReference` by the instance lock that access a static shared map